### ISSUE TITLE
cache found contacts locally as soon as they are fetched

### ISSUE THREAD

**Issue by tomholub** [FlowCrypt/flowcrypt-browser#4532]
ok. On browser extension, we actually do keep a cache and it should be a lot faster the second time. Is it not? Maybe we don't currently cache all, but only cache items that the user actually used. If that's the case, we could start caching all on browser extension too.

_Originally posted by @tomholub in https://github.com/FlowCrypt/flowcrypt-ios/issues/1741#issuecomment-1177831585_

**Comment by ioanmo226** [comment]
ok

**Comment by ioanmo226** [comment]
I think I was wrong.
As I checked, we save found google contacts in DB store.
@tomholub Please confirm if I'm correct.

https://github.com/FlowCrypt/flowcrypt-browser/blob/fae2bb7d00555a53713cb13c2300b54ff29c3aa6/extension/chrome/elements/compose-modules/compose-recipients-module.ts#L574-L576

**Comment by tomholub** [comment]
In that case, please debug the resulting behavior - maybe it's not perfect. Because it caused you to think it's not implemented. Maybe it doesn't work 100%.

**Comment by ioanmo226** [comment]
I found why current behavior let me thought it's not working.
Here's source code for search contacts and debug output for searching `e2e` string in secure compose.
As you can see, `searchContacts 7 - finishing` is called after `searchContacts 3`(find contacts from google), `searchContacts 4`(which finds contact info from sent messages).
Search progress doesn't stop on `searchContacts 2`(search from fetched db) because search result count (2) is less than MAX_CONTACTS_LENGTH(3)

In `renderSearchRes` function, we set `contacts` `display` to `none`.
And only in `searchContacts 7 - finishing` block, we show `contacts` popup by calling `renderSearchResultsLoadingDone`.

As a result, searching contacts in secure compose took long time.

```ts
private searchContacts = async (input: JQuery<HTMLElement>): Promise<void> => {
  try {
    this.contactSearchInProgress = true;
    this.view.errModule.debug(`searchContacts`);
    const substring = Str.parseEmail(String(input.val()), 'DO-NOT-VALIDATE').email;
    this.view.errModule.debug(`searchContacts.query.substring(${JSON.stringify(substring)})`);
    if (!substring) {
      this.view.errModule.debug(`searchContacts 1`);
      this.hideContacts(); // todo - show suggestions of most contacted ppl etc
      return;
    }
    const contacts: ContactPreview[] = await ContactStore.search(undefined, { substring });
    this.view.errModule.debug(`searchContacts substring: ${substring}`);
    this.view.errModule.debug(`searchContacts db count: ${contacts.length}`);
    this.renderSearchRes(input, contacts, { substring });
    if (contacts.length >= this.MAX_CONTACTS_LENGTH || !(this.canReadEmails || this.googleContactsSearchEnabled)) {
      this.view.errModule.debug(`searchContacts 2, count: ${contacts.length}`);
      return;
    }
    this.view.errModule.debug(`searchContacts 3`);
    const foundOnGoogle = await this.searchContactsOnGoogle(substring, contacts);
    await this.addApiLoadedContactsToDb(foundOnGoogle);
    contacts.push(...foundOnGoogle.map(c => ContactStore.previewObj({ email: c.email, name: c.name })));
    this.renderSearchRes(input, contacts, { substring });
    if (contacts.length >= this.MAX_CONTACTS_LENGTH) {
      this.view.errModule.debug(`searchContacts 3.b, count: ${contacts.length}`);
      return;
    }
    this.view.errModule.debug(`searchContacts 4`);
    if (this.canReadEmails && !foundOnGoogle.length) {
      this.view.errModule.debug(`searchContacts (Gmail Sent Messages) 6.b`);
      await this.guessContactsFromSentEmails(substring, contacts, async guessed => {
        await this.addApiLoadedContactsToDb(guessed.new);
        contacts.push(...guessed.new.map(c => ContactStore.previewObj({ email: c.email, name: c.name })));
        this.renderSearchRes(input, contacts, { substring });
      });
    }
  } catch (e) {
    Ui.toast(`Error searching contacts: ${ApiErr.eli5(e)}`, false, 5);
    throw e;
  } finally {
    this.view.errModule.debug('searchContacts 7 - finishing');
    this.contactSearchInProgress = false;
    this.renderSearchResultsLoadingDone();
  }

```
> [aPcnv] searchContacts
> compose-err-module.js:76 [aPcnv] searchContacts.query.substring("e2e")
> compose-err-module.js:76 [aPcnv] searchContacts substring: e2e
> compose-err-module.js:76 [aPcnv] searchContacts db count: 2
> compose-err-module.js:76 [aPcnv] renderSearchRes len: 2
> compose-err-module.js:76 [aPcnv] searchContacts 3
> compose-err-module.js:76 [aPcnv] searchContacts (Google API) 5
> compose-err-module.js:76 [aPcnv] addApiLoadedContactsToDb 1
> compose-err-module.js:76 [aPcnv] renderSearchRes len: 2
> compose-err-module.js:76 [aPcnv] searchContacts 4
> compose-err-module.js:76 [aPcnv] searchContacts (Gmail Sent Messages) 6.b
> compose-err-module.js:76 [aPcnv] guessContactsFromSentEmails start
> compose-err-module.js:76 [aPcnv] addApiLoadedContactsToDb 1
> compose-err-module.js:76 [aPcnv] renderSearchRes len: 2
> compose-err-module.js:76 [aPcnv] guessContactsFromSentEmails end
> compose-err-module.js:76 [aPcnv] searchContacts 7 - finishing
> compose-err-module.js:76 [aPcnv] ** input_to giving focus to(undefined)

**Comment by ioanmo226** [comment]
I wonder why we have to set `display` `none` to `contacts` popup in `renderSearchRes` function.

https://github.com/FlowCrypt/flowcrypt-browser/blob/f466938780a14661814ae89cac521b775cfc5513/extension/chrome/elements/compose-modules/compose-recipients-module.ts#L705

**Comment by tomholub** [comment]
@rrrooommmaaa please have a look to comment if needed

**Comment by rrrooommmaaa** [comment]
> I wonder why we have to set `display` `none` to `contacts` popup in `renderSearchRes` function.

Apparently, with this fix #3849 to call `showContacts` separately. That PR wasn't unfortunately accompanied with a unit test, so I suggest to create one so we can check for regression with this PR (issue #3711)

**Comment by ioanmo226** [comment]
I see. Thanks for your notice
But I think that's not worth testing and not sure what to test.
Do we have to test if loading spinner not shows while user types? (In live test)?
In mock test, that could be several milliseconds and even in life test, it could be less than 1 sec.

**Comment by rrrooommmaaa** [comment]
> But I think that's not worth testing and not sure what to test.

You're right, the code that previously inserted the `<li>` element containing `loading...` was completely removed, so regression isn't possible.
I can only suspect that the rendering is delayed until the full search has finished to prevent entries from "dancing" (new entries are added and the list is re-sorted).

**Comment by ioanmo226** [comment]
IMHO `dancing` would not affect user in bad way. 
As we have debounce time & api call duration, add transition would look smooth.

**Comment by rrrooommmaaa** [comment]
> IMHO `dancing` would not affect user in bad way.

Indeed, this is a rare case when the user aimed to click on a contact and the list of contacts has changed at the very last moment and the user clicked on something else, so I think we can ignore that.
Also interesting question: why do we need `contactSearchInProgress` state variable -- looks like `renderSearchRes` is only called when `contactSearchInProgress=true`, right? Hence this condition is always true: `((renderableContacts.length > 0 || this.contactSearchInProgress) || !this.googleContactsSearchEnabled)`, so we can simplify that, I guess.

**Comment by ioanmo226** [comment]
Nice catch. You are right.
Just removed that variable
